File: server/lib/viewport.js
'use strict';

const ROW_STEP = 1;
const COL_STEP = 8;

function panViewport(view, key) {
  const { row, col, follow } = view;
  if (key === 'up') return { row: Math.max(0, row - ROW_STEP), col, follow: false };
  if (key === 'down') return { row: row + ROW_STEP, col, follow: false };
  if (key === 'left') return { row, col: Math.max(0, col - COL_STEP), follow };
  if (key === 'right') return { row, col: col + COL_STEP, follow };
  return view;
}

function anchorRow(selRow, viewRows) {
  return Math.max(0, selRow - (viewRows - 2));
}

function windowLines(grid, view, dims) {
  const out = [];
  for (let r = view.row; r < view.row + dims.rows && r < grid.length; r += 1) {
    const ln = grid[r];
    out.push({ text: String(ln.text).slice(view.col, view.col + dims.cols), color: ln.color });
  }
  return out;
}

module.exports = { panViewport, windowLines, anchorRow, ROW_STEP, COL_STEP };

File: server-go/internal/screen/viewport.go
package screen

const (
	RowStep = 1
	ColStep = 8
)

type View struct {
	Row    int
	Col    int
	Follow bool
	SelRow int
}

func PanViewport(v View, key string) View {
	switch key {
	case "up":
		return View{Row: max(0, v.Row-RowStep), Col: v.Col, Follow: false, SelRow: v.SelRow}
	case "down":
		return View{Row: v.Row + RowStep, Col: v.Col, Follow: false, SelRow: v.SelRow}
	case "left":
		return View{Row: v.Row, Col: max(0, v.Col-ColStep), Follow: v.Follow, SelRow: v.SelRow}
	case "right":
		return View{Row: v.Row, Col: v.Col + ColStep, Follow: v.Follow, SelRow: v.SelRow}
	}
	return v
}

func AnchorRow(selRow, viewRows int) int {
	return max(0, selRow-(viewRows-2))
}

func WindowLines(grid []Line, v View, rows, cols int) []Line {
	out := []Line{}
	for r := v.Row; r < v.Row+rows && r < len(grid); r++ {
		runes := []rune(grid[r].Text)
		lo := min(v.Col, len(runes))
		hi := min(v.Col+cols, len(runes))
		out = append(out, Line{Text: string(runes[lo:hi]), Color: grid[r].Color})
	}
	return out
}
